shell/inject.1
.\" Process this file with
.\" groff -man -Tascii inject.1
.\"
.TH TWOPENCE_INJECT "1" "@DATE@" "Twopence @VERSION@" "User Commands"

.SH NAME
twopence_inject \- inject a test file

.SH SYNOPSIS
.B twopence_inject [
.I OPTION
.B ]... 
.I TARGET
.B  
.I LOCAL
.B  
.I REMOTE

.SH DESCRIPTION
.B twopence_inject
injects a local file onto the system under test (SUT).
.PP
The SUT can be accessed by any means provided by Twopence:
virtio (for QEmu/KVM virtual machines), ssh, or serial
(with a null-modem cable). This access method and the exact
location of the SUT are defined by the
.I TARGET
parameter. With the virtio and serial methods, the program
.B twopence_test_server
must be installed and running on the SUT. With the ssh method,
a SSH server must be installed and running on the SUT.
.PP
The transfer can be done as any local user on the SUT (default is
.I root
user). There is absolutely no access control, so make sure you
use Twopence only on test machines on which you have exclusive
access.

.SH OPTIONS
.IP \fB\-u\fR\ \fIUSERNAME\fR
.IP\fB\-\-user\fR=\fIUSERNAME\fR
Define the username under which the file is written
on the system under test.
.IP \fB\-v\fR
.IP \fB\-\-version\fR
Display version information.
.IP \fB\-h\fR
.IP \fB\-\-help\fR
Display a short error message.
.PP
.I TARGET
obeys the following syntax:
.PP
.IP \fBvirtio:\fR\fISOCKETNAME\fR
for the access through virtio to
a QEmu/KVM virtual machine. \fISOCKETNAME\fR is the filename of
the UNIX domain socket used by the virtio channel. It can be defined
for example with help of
.BR virt-manager (5)
command.
.PP
.IP \fBssh:\fR\fIHOSTNAME\fR|\fIIPADDRESS\fR\fB[:\fR\fIPORTNUMBER\fR\fB]\fR
for the access though SSH to a remote host.
.PP
.IP \fBserial:\fR\fIDEVICENAME\fR
for the access through a serial port. \fIDEVICENAME\fR is the filename
of a UNIX character device describing the serial port.
.PP
.I LOCAL
is the filename of the local file to send.
.PP
.I REMOTE
is the filename under which the file should be saved on the SUT.

.SH EXAMPLES
Example syntax for the virtio access method:
.IP \fBtwopence_inject\ \-u\ johndoe\ virtio:/tmp/sut.sock\ local_file.txt\ remote_file.txt\fR
it will use
.I /tmp/sut.sock
to communicate with the QEmu/KVM host.
.PP
Example syntax for the ssh access method:
.IP \fBtwopence_inject\ \-\-user=johndoe\ ssh:host.example.com\ local_file.txt\ remote_file.txt\fR
it is functionally equivalent to
\fBscp local_file.txt johndoe@host.example.com:remote_file.txt\fR
without server footprint verification.
.PP
Example syntax for the serial access method:
.IP \fBtwopence_inject\ serial:/dev/ttyS0\ local_file.txt\ remote_file.txt\fR

.SH BUGS
Do not expect any security from Twopence. Use with care. You've been warned.
.PP
For the moment, only one session with the remote host can be used at
a time.
.PP
For the moment, it is not possible to use jokers like * or ? to transfer
several files at once. It is also not possible to transfer recursively a
whole directory.

.SH AUTHOR
The Twopence developpers at SUSE Linux.

.SH SEE ALSO
.BR twopence_command (1),
.BR twopence_extract (1),
.BR twopence_exit (1),
other shell commands to access the System Under Test.
.PP
.BR twopence_test_server (1),
the server to be installed on the System Under Test when using
the virtio or serial access methods.


shell/inject.c
/*
File injection command. It is used to send a file into some testing environment.
Currently the only supported environment is a livirt virtual machine.


Copyright (C) 2014-2015 SUSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 2.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>

#include "shell.h"
#include "twopence.h"
#include "version.h"

char *short_options = "u:dvh";
struct option long_options[] = {
  { "user", 1, NULL, 'u' },
  { "debug", 0, NULL, 'd' },
  { "version", 0, NULL, 'v' },
  { "help", 0, NULL, 'h' },
  { NULL, 0, NULL, 0 }
};

// Display a message about the command usage
void usage(const char *program_name)
{
    fprintf(stderr, "Usage: %s [<options>] <target> <local file> <remote file>\n\
Options: -u|--user <user>: user injecting the file (default: root)\n\
         -d|--debug: print debugging information\n\
         -v|--version: print version information\n\
         -h|--help: print this help message\n\
Target: serial:<character device>\n\
        ssh:<address and port>\n\
        virtio:<socket file>\n", program_name);
}

// Main program
int main(int argc, char *argv[])
{
  int option;
  const char *opt_user,
             *opt_target, *opt_local, *opt_remote;
  struct twopence_target *target;
  int rc, remote_error;

  // Parse options
  opt_user = NULL;
  while ((option = getopt_long(argc, argv, short_options, long_options, NULL))
         != -1) switch(option)         // parse individual options
  {
    case 'u': opt_user = optarg;
              break;
    case 'd': twopence_debug_level++;
	      break;
    case 'v': printf("%s version %s\n", argv[0], TWOPENCE_VERSION);
              exit(RC_OK);
    case 'h': usage(argv[0]);
              exit(RC_OK);
    default: usage(argv[0]);
             exit(RC_INVALID_PARAMETERS);
  }
  if (opt_user == NULL)                // default user
    opt_user = "root";
  if (argc != optind + 3)              // mandatory arguments: target, local and remote
  {
    usage(argv[0]);
    exit(RC_INVALID_PARAMETERS);
  }
  opt_target = argv[optind++];
  opt_local = argv[optind++];
  opt_remote = argv[optind++];

  rc = twopence_target_new(opt_target, &target);
  if (rc < 0)
  {
    twopence_perror("Error while initializing library", rc);
    exit(RC_LIBRARY_INIT_ERROR);
  }

  // Inject file
  rc = twopence_inject_file
         (target, opt_user, opt_local, opt_remote,
          &remote_error, true);
  if (rc == 0)
    printf("File successfully injected\n");
  else
  {
    twopence_perror("Unable to inject file", rc);
    rc = RC_EXTRACT_FILE_ERROR;
  }
  if (remote_error != 0)
    fprintf(stderr, "Remote error code: %d\n", remote_error);

  // End library
  twopence_target_free(target);
  return rc;
}
